### PR TITLE
Slice generation with yt compliance for viz

### PR DESCRIPTION
**Pull request by RevathiJambunathan**
Added yt visualization capability for slice diagnostics in WriteSlicePlotFile ().
The raw field data is written out in diags/slice_plotfiles/pltXXXXX based on the interval given in the input file.

### CODE REVIEW

**Comment by LDAmorim** [comment]
Hi Revathi,

Thank you for making it easy to read this with yt!
Could you compiled this branch with openmp?
In 3D I think it is useful to use it.
But I could only compile with USE_OMP=False (your other PR allowed OMP).
It might be that I installed something wrong in my computer too.

Thanks,
Diana

**Comment by RevathiJambunathan** [comment]
Hi Diana,

Thank you for testing this branch. 
Yes I had compiled with USE_OMP=TRUE for both 2D and 3D.

**Comment by LDAmorim** [comment]
Hi Revathi,
I will make some tests then to understand why I had the error.
Thanks!
Diana

**Comment by MaxThevenet** [comment]
@RevathiJambunathan Thanks for this PR! 
Is it already tested and ready for review?

**Comment by RevathiJambunathan** [comment]
Hi Maxence, 

I still have to merge it with the changes made in the Previous PR (PR #97 ), since I added the moving window capability in the previous PR. 
I thought once the changes there are  accepted, I will merge it and update this PR. 

Otherwise -- the output format has been tested -- but it wont work for a moving window velocity > 1 in its current state. 

Should I merge the previous PR right away?

Thank you
Revathi

**Comment by MaxThevenet** [comment]
Tests failed after 45s, not sure why this happened. I re-built them, I'll review the PR when they pass.

**Comment by LDAmorim** [comment]
Hi Revathi,

I was just testing your new branch with the LWFA 3D example, which has `warpx.moving_window_v = 1.0` (the same that I sent you the other week).

Running this new PR, I still got similar results as with the previous version of it.
Is the option of moving window with v>0, not working?

When I try to write the full 3D dataset on the slice, using the following section in the input file:

```
slice.dom_lo = -30.0e-6  -30.0e-6 -56.0e-6
slice.dom_hi =  30.0e-6 30.0e-6 12.0e-6
slice.coarsening_ratio = 1 1 1
slice.plot_int = 100
```

And try to open it in yt with:
```
ds = yt.load(file)
ad = ds.all_data()
f_test= ad['Ey_slice']
print(f_test.shape)
```

I get (262144, 4) as the shape of the slice information.
In this case 262144 is the total number of cells (32*32*256).
But I still don't understand what the 4 columns stand for.
Also, why is the slice not a 3D array of (32,32,256) shape?

How do you open the slice with the correct shape using yt?

Trying to open it resorting to `ds.covering_grid(level=0,left_edge=ds.domain_left_edge, dims=ds.domain_dimensions)` still gives the error: `Buffer has wrong number of dimensions (expected 1, got 2)`.

When I test the same input, but writing only 2D information in the slice:
```
slice.dom_lo = 0.0 -30.0e-6 -56.0e-6
slice.dom_hi = 0.0 30.0e-6 12.0e-6
```

I get as an output of the same commands (16384, 4).
Which makes it seem like there are two times the size of all cells in one slice (32*256).
How should I interpret that data?

It would be nice to also test the boosted frame (which needs moving window at v=+c) and dumping 1D and 2D slices at any desired location and 3D slices.

Feel free to write back if you have any questions or if I can help,
Diana

**Comment by RevathiJambunathan** [comment]
Hi Diana, 

Thank you for testing the yt compliant output file. 

I have modified the output format such that the ghost cells are not printed. 
Now I can visualize the data on yt. 

Here are the code snippets for visualizing the slice. Note that I ran the code such that the slice is essentially a copy of the underlying domain, similar to what we discussed offline. 

import yt
ds = yt.load("plt00200")
ds.field_list
slc = yt.SlicePlot(ds, "z", "Ez_slice",center=[0., 0., 0.])
slc

I talked with Andrew about the issue of printing the field value at a certain point in the domain. 
It looks like the current version of yt does not support probing values of non cell-centered data at a given grid point. Andrew has said he will submit a PR for this on the yt repository. 

So to summarize, the slice diagnostic works for a moving grid and can be visualized using yt, but the values cannot be probed yet. 
Please let me know if you have any questions or suggestions as you try this feature for other cases. 

Thank you
Revathi

**Comment by LDAmorim** [comment]
Hi Revathi,

Thanks! These are great news.

I don’t think I understand what you wrote regarding the none cell-centered data.

Would it not be possible to determine in which cell a value, set by the user in the input file, lies?
Then you could store the slice data of that cell.

I will test this for the LWFA lab / boosted frames example runs as soon as I can.

Thanks,
Diana

**Comment by atmyers** [comment]
Hi Diana,

Specifically, the `covering_grid` function that we sometimes use for WarpX doesn't work with non-cell centered data. I know how to fix this and will make a pull request to yt. In the meantime it would be still be possible to figure out which grid contains a point, access that grid, and then get the field values, but that is a bit more manual.

**Comment by LDAmorim** [comment]
Hi Andrew,

Oh I see. That makes sense.

Thank you

**Comment by RevathiJambunathan** [comment]
Hi Diana, 

Also, I have not included boosted frame capability for the slice diagnostic yet --
Will do that in a separate PR. 

Currently it only has the moving window capability. 

-Reva

File: Source/Diagnostics/WarpXIO.cpp [hunk 1]
```diff
@@ -746,29 +746,72 @@ WarpX::WriteJobInfo (const std::string& dir) const
 
 
 /* \brief
- *  The slice is ouput using visMF and can be visualized used amrvis. 
+ *  The raw slice data is written out in the plotfile format and can be visualized using yt. *  The slice data is written to diags/slice_plotfiles/pltXXXXX at the plotting interval.
```
Comment (MaxThevenet) [inline Source/Diagnostics/WarpXIO.cpp]: Looks like there's a line break missing here.

File: Source/Diagnostics/WarpXIO.cpp [hunk 2]
```diff
@@ -746,29 +746,72 @@ WarpX::WriteJobInfo (const std::string& dir) const
...
  */
 void
 WarpX::WriteSlicePlotFile () const
 {
-    if (F_fp[0] ) {
-       VisMF::Write( (*F_slice[0]), "vismf_F_slice");
-    }
+    // writing plotfile //
+    const std::string& slice_plotfilename = amrex::Concatenate(slice_plot_file,istep[0]);
+    amrex::Print() << " Writing slice plotfile " << slice_plotfilename << "\n";
```
Comment (MaxThevenet) [inline Source/Diagnostics/WarpXIO.cpp]: Can we remove this print statement?

File: Source/Diagnostics/WarpXIO.cpp [hunk 3]
```diff
@@ -746,29 +746,72 @@ WarpX::WriteJobInfo (const std::string& dir) const
...
+                                   "Level_", "Cell", rfs);
+
+    for (int lev = 0; lev < nlevels; ++lev) 
+    {
+        const std::unique_ptr<MultiFab> empty_ptr;
+        const std::string raw_spltname = slice_plotfilename + "/raw_fields";
+        amrex::Print() << " raw spltname " << raw_spltname << "\n";
+        const DistributionMapping &dm = Efield_slice[lev][0]->DistributionMap();
+ 
+        WriteRawField( *Efield_slice[lev][0], dm, raw_spltname, level_prefix, "Ex_slice", lev, 0);
```
Comment (MaxThevenet) [inline Source/Diagnostics/WarpXIO.cpp]: Why is it required to write slice data as a raw fields?

File: Source/Diagnostics/WarpXIO.cpp (hunk 1 above)
Comment (RevathiJambunathan) [reply Source/Diagnostics/WarpXIO.cpp]: Thank you for catching that. Fixed it now.

File: Source/Diagnostics/WarpXIO.cpp (hunk 2 above)
Comment (RevathiJambunathan) [reply Source/Diagnostics/WarpXIO.cpp]: Done.

File: Source/Diagnostics/WarpXIO.cpp (hunk 3 above)
Comment (RevathiJambunathan) [reply Source/Diagnostics/WarpXIO.cpp]: The slice data inherits the index type of the underlying data. 
As a result, we just write out the raw data without converting it to cell-centered values and then writing it as a plotfile.

File: Source/Diagnostics/WarpXIO.cpp (hunk 1 above)
Comment (RevathiJambunathan) [reply Source/Diagnostics/WarpXIO.cpp]: Thank you Maxence for reviewing this PR. This branch has only been tested without Mesh Refinement. The slice generation currently works with one level only and therefore has not been tested with Mesh refinement. I have tested the slice generation for 1D 2D and 3D slices and visualized it using amrvis. I have tested visualization with 3D using yt. I will update here after I test yt-visualization with 2D and 1D configurations. I dont think I have changed the default input parameters for sub-sampling.

**Comment by MaxThevenet** [comment]
Thanks for this PR!
